### DISCUSSION TITLE
Track last valid input state and auto-reset on blur

### DISCUSSION THREAD

**Discussion by snoozbuster** [formkit/formkit#927] [General]
Hey there. I have a requirement as described in the title - I want to automatically reset an input to the last valid value when the field is blurred, if the field is currently invalid. I figured I could do this with a plugin, and this is what I cooked up:

```ts
function resetInvalidPlugin(node: FormKitNode) {
  if (node.props.type === 'group') {
    return;
  }

  let lastValue;
  let nextValue;
  let isValid;

  node.on('input', ({ payload }) => {
    // track unvalidated next value
    nextValue = payload;
  });
  node.on('unsettled:blocking', () => {
    // keep track of the fact that the field is invalid
    isValid = false;
  });
  node.on('settled:blocking', () => {
    // once the field becomes valid, note so and set the last noted input valid as the last valid value
    isValid = true;
    lastValue = nextValue;
  });
  node.on('created', () => {
    lastValue = node._value;
    const blur = node.context!.handlers.blur;
    node.context!.handlers.blur = () => {
      blur();
      if (!isValid) {
       // on blur, if the field is invalid, reset to the last valid value
        node.input(lastValue);
      }
    };
  });
}
```

This... almost works. But sometimes if the user types very quickly, an invalid value gets stored in the `lastValue` before the validation kicks in. indeed, from my logs, it seems like the blocking counter settles even while there are validation errors:

![image](https://github.com/formkit/formkit/assets/5778364/274cc9ed-3704-429e-b907-25024be452ac)

The third line here is incorrect - there is a validation rule of `['matches', '/.+\n+.+/']` on this field (it is only valid if there are at least 2 lines). it does eventually notice on lines 4 right afterward that the field is invalid, but by that time it is too late - the invalid value has already made it into the `lastValue` variable and now the resetting won't work.

Is there a better way to achieve this? I tried also using `settled:errors` as in [this example](https://formkit.com/playground?fkv=latest&fileTab=formkit.config.js&files=jc%5B%28%27name%21%27Playground.vue%60%3CscripDsetup%3E5GESM+importzarZsupported+from+URLsHhttps%3A%2F%2Fcdn...%7D5Gazwell+azsecondary+playground+filesH.%2FOtherFile.vue%7D.5consDlogX+console.log5%24script%3E55%3C%2B3%3C%25t0type%3D7form70%40submit-invalid%3D7log753N0142014ed703validation%3D7requir2014201420142014201420142014201420142014201420142014ed70%2F%3E053%24%25t%3E5%24%2B5%3CstylZscoped%3E5%2F*5vanilla+CSS+caQgo+here.5Keep+stylezscoped+to+avoid+multiplZfiles5overwriting+each+other+iQthZrender+output.5*%2F5%24style%3E5%27%3B%2C%28%27name%21%27formkit.config.js%60%2F**5+*+A+littlZplugiQthaDautomatically+_zto+thZfirsDerror.5+**%2F5%236%5E9%2853ifHJtypeX%3D%3D+%22form%229%280%23_To%5E9%2803consDelX+document.getElementById%7BJid%7D03ifHel9%28033el._IntoView%7B%7D03%290%2900%236H9%2803node.walk%7BchildX%3E+%28033GCheck+if+thizchild+haz%26033ifHYblocking%229%7C%7C+Y%26%22%7D9%280333GWZfound+aQinpuDwith+validatioQ%260333_To%7Bchild%7D0333GStop+searching033%3F033%2903%29%2C+true%7D0%2900consD8X+J80J8XH9%3D%3E+%28038%5E%7D036%7B%7D0%290node.on%7B%22unsettled%3A%26%22%2C+6%7D53%295%3F5%29555exporDdefaulD%2853plugins%3A+%5B6%5D5%295%27%7Eremovable%21true%7Esilent%21false%3B%5D053313label%3D7%25DInput703help%3D7ediDmZt2ed70%2FN3++4o+geDstart5%5Cn6_ToErrors7%5C%278onSubmitInvalid9%7D+Dt+G%2F%2F+H+%7BJnode.props.N%3E0%3C%25t03type%3D7text7Qn+X+%3DYchild.ledger.value%7B%22Ze+_scrollzs+%23functioQ%24%3C%2F%25FormKi%26errors%2Btemplate%3E5%3B%7Eadded%21true%29%3F3returQfalse%5E%7Bnode%60%27%7Eeditor%21%27%01%60%5E%3F%3B%2B%26%25%24%23z_ZYXQNJHGD9876543210_&imports=jc%28%27name%21%27ImportMap%27%7Eeditor%21%27%28*+1vue%5C%211https%3A%2F%2Fcdn.jsdelivr.net%2Fnpm%2Fvue%403%2Fdist%2Fvue.esm-browser.min.js0*%29*%27%29*%5Cn0%5C%271+0%0110*_&css-framework=genesis) but that event never seemed to be fired for me.

**Comment by fenilli** [comment]
the reason for the `settled:errors` not firing is because those are for errors and not validation errors, so errors created by the backend and added manually using `setErrors`.

As for your plugin reseting a value back to before it was invalid: https://formkit.link/edfa6f1951870b3f585346eefca6cd5d

**Reply by snoozbuster** [reply]
ah, perfect! thank you!
